src/nvim/os/stream.c

```diff
@@ -0,0 +1,26 @@
+// Functions for working with stdio streams (as opposed to RStream/WStream).
+
+#include <stdio.h>
+#include <stdbool.h>
+
+#include <uv.h>
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "os/stream.c.generated.h"
+#endif
+
+/// Sets the stream associated with `fd` to "blocking" mode.
+///
+/// @return `0` on success, or `-errno` on failure.
+int stream_set_blocking(int fd, bool blocking)
+{
+  int flags = fcntl(fd, F_GETFL, 0);
+  int err = 0;
+  if (!blocking && !(flags & O_NONBLOCK)) {
+    err = fcntl(fd, F_SETFL, flags | O_NONBLOCK);
+  } else if (blocking && (flags & O_NONBLOCK)) {
+    err = fcntl(fd, F_SETFL, flags & ~O_NONBLOCK);
```

Comment:
Here I'm calling `fcntl` instead of the previous libuv-based implementation (see below for reference), because in the case where some component other than `main.c` was setting `stderr` to non-blocking, using the libuv-based implementation would result in this runtime assertion:

    .deps/build/src/libuv/src/unix/core.c:833: uv__io_stop: Assertion `loop->watchers[w->fd] == w' failed.

**As best as I can guess,** this is some sort of conflict with an existing handle in libuv.

---

      int stream_set_blocking(int fd, bool blocking)
      {
        uv_pipe_t stream;
        uv_pipe_init(uv_default_loop(), &stream, 0);
        uv_pipe_open(&stream, fd);
        int retval = uv_stream_set_blocking((uv_stream_t *)&stream, blocking);
        uv_close((uv_handle_t *)&stream, NULL);
        return retval;
      }

---

I'd much prefer to use the libuv-based solution, but this is a temporary band-aid to fix an important bug. There's a remote possibility that it's a libuv bug: https://github.com/joyent/libuv/issues/838#issuecomment-58621040 

More likely it is related to use of a watcher on the same `fd`, but the watcher is not being stopped before `stream_set_blocking()` is called. More details: https://github.com/yfinkelstein/node-zookeeper/pull/101